tfhe/src/shortint/server_key/tests/noise_level.rs

```diff
@@ -0,0 +1,288 @@
+        NoiseLevel::NOMINAL * 2
+    });
+    test_fn(
+        &|sk, ct1, ct2| {
+            ServerKey::unchecked_evaluate_bivariate_function_assign(sk, ct1, ct2, |_, _| 0)
+        },
+        &|_, _| NoiseLevel::NOMINAL,
+    );
+}
+
+fn test_ct_scalar_op_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext, scalar: u8) {
+    let test_fn = |op: &dyn Fn(&ServerKey, &Ciphertext, u8) -> Ciphertext,
+                   predicate: &dyn Fn(NoiseLevel, u8) -> NoiseLevel| {
+        assert!(op(sk, ct, scalar).noise_level() == predicate(ct.noise_level(), scalar));
+    };
+
+    test_fn(&ServerKey::unchecked_scalar_add, &|ct_noise, _| ct_noise);
+    test_fn(&ServerKey::unchecked_scalar_sub, &|ct_noise, _| ct_noise);
+    test_fn(&ServerKey::unchecked_scalar_mul, &|ct_noise, scalar| {
+        ct_noise * scalar as usize
+    });
+    if scalar != 0 {
+        test_fn(&ServerKey::unchecked_scalar_div, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+        test_fn(&ServerKey::unchecked_scalar_mod, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+    }
+    test_fn(&ServerKey::unchecked_scalar_bitand, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitor, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitxor, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    if scalar < 8 {
+        test_fn(
+            &ServerKey::unchecked_scalar_left_shift,
+            &|ct_noise, scalar| ct_noise * (1 << scalar as usize),
+        );
+    }
+    test_fn(&ServerKey::unchecked_scalar_right_shift, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+}
+
+fn test_ct_scalar_op_assign_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext, scalar: u8) {
+    let test_fn = |op: &dyn Fn(&ServerKey, &mut Ciphertext, u8),
+                   predicate: &dyn Fn(NoiseLevel, u8) -> NoiseLevel| {
+        let mut clone = ct.clone();
+        op(sk, &mut clone, scalar);
+        assert!(clone.noise_level() == predicate(ct.noise_level(), scalar));
+    };
+
+    test_fn(&ServerKey::unchecked_scalar_add_assign, &|ct_noise, _| {
+        ct_noise
+    });
+    test_fn(&ServerKey::unchecked_scalar_sub_assign, &|ct_noise, _| {
+        ct_noise
+    });
+    test_fn(
+        &ServerKey::unchecked_scalar_mul_assign,
+        &|ct_noise, scalar| ct_noise * scalar as usize,
+    );
+    if scalar != 0 {
+        test_fn(&ServerKey::unchecked_scalar_div_assign, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+        test_fn(&ServerKey::unchecked_scalar_mod_assign, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+    }
+    test_fn(&ServerKey::unchecked_scalar_bitand_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitor_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitxor_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    if scalar < 8 {
```

Comment:
why this special case ?

Reply:
To avoid an overflow in `let scalar = 1_u8 << shift;`

Reply:
ok thanks